tobias/tools/submerge.py

```diff
@@ -0,0 +1,164 @@
+
+
+def run_submerge(args):
+
+    logger = TobiasLogger("SubMerge", args.verbosity)
+    logger.begin()
+    parser = add_submerge_arguments(argparse.ArgumentParser())
+    logger.arguments_overview(parser, args)
+
+    outdir = os.path.dirname(args.output)
+
+    # Get all TFBS files
+    tfbs_files = glob.glob(os.path.join(args.tfbs, "*/*_overview.txt"))
+
+    # Read regions
+    logger.debug("Reading regions from: " + args.regions)
+    checked_regions = []
+    with open(args.regions) as f:
+        for i, line in enumerate(f):
+            splitline = line.strip().split("\t")
+
+            # if more than 6 columns, remove the last columns
+            if len(splitline) > 6:
+                splitline = splitline[:6]
+
+            # check if all necessary columns are present
+            while len(splitline) < 6:
+                match len(splitline):
+                    case 3:
+                        splitline.append(f'region_{i}')  # add a unique identifier
+                    case 4:
+                        splitline.append('.')            # add a score placeholder
+                    case 5:
+                        splitline.append('.')            # add a strand placeholder
+
+            # if strand was not specified, intersect with both
+            if splitline[5] != '+' and splitline[5] != '-':
+
+                splitline[5] = '+'
+                line_string = "\t".join(splitline)
+                checked_regions.append(line_string)
+
+                splitline[5] = '-'
+                line_string = "\t".join(splitline)
+                checked_regions.append(line_string)
+
+            else:
+                line_string = "\t".join(splitline)
+                checked_regions.append(line_string)
+
+    # write checked regions to file
+    logger.debug("Writing checked regions to file")
+    regions = os.path.join(outdir, "regions.tmp")
+    with open(regions, "w") as f:
+        for line in checked_regions:
+            f.write(line + "\n")
+
+    # Read TFs
+    if args.TF is not None:
+        logger.debug("Reading TFs from: " + args.TF)
+        tfs = pd.read_csv(args.TF, header=None, names=["TF"])
+        tfs = tfs["TF"].tolist()
+
+        # get only those TFs that were provided
+        tfbs_files = [tfbs for tfbs in tfbs_files if os.path.basename(os.path.dirname(tfbs)) in tfs]
+
+    # intersect each TFBS file with the regions
+    logger.debug('Intersecting query with all TBFS files')
+    all_intersections = ''
+    for file in tfbs_files:
+        tf = os.path.basename(os.path.dirname(file))
+
+        # remove header
+        headless_file = os.path.join(outdir, f"{tf}.headless")
+        command = f'sed "1d" {file} > {headless_file}'
+        subprocess.run(command, shell=True)
+
+        # intersect
+        command = f"bedtools intersect -a {regions} -b {headless_file} -wa -wb"
+        intersection = subprocess.check_output(command, shell=True).decode("utf-8")
+        all_intersections += intersection
```

Comment:
use pybedtools

Reply:
I did it this way because pybedtools creates a temp file for every action, and since I just want to take the output and append to my other outputs instantly, I thought it might be more better to avoid that. But if you insist that pybedtools should be used here because there is some other advantage, I can change that.

Reply:
That is a good point. Usually, I would try to use python code as much as possible and use an interface as pybedtools to handle calling tools. This is usually more robust and easier to debug. In this case, I would recommend a try-finally construct to remove temp files even when an error occurs ([pybedtools temp principle](https://daler.github.io/pybedtools/topical-design-principles.html#temp-principle)). But not creating a temp file might be better in this case. It probably comes down to speed vs. robustness. But it should be fine either way so feel free to decide.